### ISSUE TITLE
It's not serving webp image anymore after Rocket-Nginx.

### ISSUE THREAD

**Issue by Alnaggar94** [SatelliteWP/rocket-nginx#136]
**Describe the bug**
Hello, everything works fine except the "webp" images, I'm using (WebP Express plugin for webp images) it's not serving webp image anymore after Rocket-Nginx.

**Versions**
What version of Nginx are you using ?
latest
What version of Rocket-Nginx are you using ?
latest
What version of WP Rocket are you using ?
latest
Are you using Nginx as a reverse proxy (with Apache for instance) ?
No, only Nginx

**Comment by mathieu-aubin** [comment]
'latest' only is relevant for a very short period of time. Don't forget that theses issues might be helpful for future generations and that by adding theses simple details, it would make it easier to go back in time and research/recreate your issue in... let's say... 45 decades from now.... Just a thought

**Comment by maximejobin** [comment]
WebP is not supported yet. That is why.

I'm working on a new version what will support it.

**Comment by amityweb** [comment]
Can we not just add this into the location section in the rocket-nginx.tmpl file? This is the code from Imagify (same developer as WP Rocket)

```
    add_header Vary Accept;
    if ($http_accept ~* "webp"){
        set $imwebp A;
    }
    if (-f $request_filename.webp) {
        set $imwebp  "${imwebp}B";
    }
    if ($imwebp = AB) {
        rewrite ^(.*) $1.webp;
    }
```

**Comment by shaneonabike** [comment]
This is awesome!

**Comment by mdotk** [comment]
@amityweb that didn't work for me sadly

**Comment by amityweb** [comment]
I just checked and still working for me. 

I did have trouble initially getting it to work in Nginx. This is because I initially added it AFTER another directive which already used this location. I noticed because I had low expire headers. Turned out my host set these already, and in Nginx you can't override a location, it sets the first one. 

I had to incldue the rocket Nginx file before the other location directive set in Nginx by my host, then it worked.

And of course, the webp images do need to exist. 

I use RunCloud to manage the server, and RunCloud have a facility to add directives in various locations so I just had to add rocket nginx in the location that comes before the file where the above location ~* was set. 

Hope all that makes sense!

**Comment by mathieu-aubin** [comment]
Will most likely fail with services such as CloudFlare and heavy-ish caching/page rules settings

**Comment by maximejobin** [comment]
Currently waiting on a status from the WP Rocket team:
https://github.com/wp-media/wp-rocket/issues/3207

**Comment by shaneonabike** [comment]
Just to be clear @amityweb how did you get this to work? I added it but am still getting headers indicating that they can't be found. i _assume_ for the time being you didn't checkout of the WebP compatibility in WP Rocket is that right? I'd love to integrate this!

**Comment by amityweb** [comment]
the default.conf hfile generated as the following code:
```
# BROWSER MEDIA CACHE
#
location ~* \.(#!# EXTENSION_MEDIAS #!#)$ {
	etag on;
	expires 30d;
	#!# HEADER_MEDIAS #!#
}
```

This means because location ~* is set here, I cant set it anywhere else in my own Nginx conf files. You can only set it once and it uses that, so I cant then set it to change or add in my own directives.

Therefore I have had to modify this default.conf as follows. I did add it into the tmpl file too, in case its regenerated, I cant recall if that works. 

Note the change to expires 30d; too as Google marks it down for 30 days now.

```
# BROWSER MEDIA CACHE
#
location ~* \.(ico|gif|jpe?g|png|svg|eot|otf|woff|woff2|ttf|ogg)$ {
	etag on;
	expires 120d;
	
	add_header Vary Accept;
    if ($http_accept ~* "webp"){
        set $imwebp A;
    }
    if (-f $request_filename.webp) {
        set $imwebp  "${imwebp}B";
    }
    if ($imwebp = AB) {
        rewrite ^(.*) $1.webp;
    }
}
```

**Comment by shaneonabike** [comment]
Cool that's what I did but I was having issues whereby this config was trying to find index.html but that doesn't exist because I had set WebP in WP Rocket. I was wondering if you had set that in settings.

**Comment by amityweb** [comment]
I have not seen that code, I just used what I put above and its working for me, that fixed my issue right away of webp nort being served. 

I sone use WP Rocket to set Webp though, I use Imagify, as that created the Webp too, but I cant see that setting affecting it, as its controlled in Nginx itself. WP Rocket or Imagify would add htaccess code but thats not used with Nginx for images. 

I use RunCloud to manage the servers, and they have their own Nginx setup, so maybe there is something else in Nginx affecting yours. RunCloud has a tool to add in whatever Nginx configs you want, and for various locations, but none of them worked for Webp due to Rocket Nginx already being loaded before them all. So my edit worked for me.

**Comment by Alnaggar94** [comment]
@amityweb - Thank you so much, your code solved the problem, I really appreciate your help.

**Comment by Alnaggar94** [comment]
@amityweb - Acualy after some tests I found out it works and doesn't work at the same time! !!? if you add the code you provided in the default.conf file when you preloaded the cache and activate the setting in WP Rocket (Enable Webp Caching) it will work but once you clear the cache and preload the cache again it will not work as the cache files in ( /sites/purelaza.org/htdocs/wp-content/cache/wp-rocket/domain.com ) will be generated with"webp" at the end of every file name while rocket nginx expects it to be without "webp" see the screenshot. How can we solve this? Does anybody have a solution?

@maximejobin - any idea how can we solve this?


![ppppppp](https://user-images.githubusercontent.com/49800146/102923962-b90d1d80-44a1-11eb-81ea-39b177142e50.JPG)

**Comment by Alnaggar94** [comment]
@amityweb  Actually after some tests, I found out it works and doesn't work at the same time! !!? If you add the code to default.conf and activate the setting in WP Rocket (Enable Webp Caching) then clear the cache and preloaded again the cache will stop working, and if you disable the "Enable Webp Caching" the cache work but not serving webp , as the cache files in ( /sites/purelaza.org/htdocs/wp-content/cache/wp-rocket/domain.com ) will be generated with"webp" at the end of every file name while rocket Nginx expects it to be without "webp" see the screenshot. How can we solve this? Does anybody have a solution?

@maximejobin - any idea how can we solve this?

![ppppppp](https://user-images.githubusercontent.com/49800146/102926652-73068880-44a6-11eb-858c-d6f029d45c2a.JPG)

**Comment by amityweb** [comment]
I dont think I am having this issue, I have not received any reports of issues on multiple websites using this for months. I cannot see this same issue or headers 

I am not enabling Webp cache n WP Rocket though. I use Imagify to auto generate the Webp, then normally on Apache Imagify adds the webp code to .htaccess but in Nginx case we have to manually add it as above. Imagify is made by the WP Rocket team though. 

So maybe there is a difference between caching webp in WP Rocket, or using Imagify and the above code. 

You dont need WP Rocket to cache images in its cache folder though do you? Its not dynamic content. It wont serve images any faster the first time it loads. Then the image is cached in the users browser. So doesn't matter where the image is, in a cache folder or what not on the server. Maybe disable the webp caching, see what happens. As long as you (1) can generate the Webp images (Manual or Imagify or some other) and (2) Have the caching in Ngnix to tell users browsers how long to cache for, then all should be OK?

**Comment by Alnaggar94** [comment]
@amityweb - Thank you so much for your time, I just realized that it works fine, I get confused that the image URL not ending with .webp and just found out that it doesn't have to be that way as it appears in the Inspect Tool as webp type. Browsers looks at the content type header rather than the URL to determine what it is that it gets. I'm using it with WebP Express plugin and everything works great.

![webp-rewrite-rules-guide-e1566400612987](https://user-images.githubusercontent.com/49800146/102995750-c3c3c300-4532-11eb-9980-6cf5d266595f.png)

**Comment by shaneonabike** [comment]
On a side "topic" I am just curious if folks have this tested on Safari 13. I just ran into an issue whereby the webpage won't load properly because webp files are being served. It could be my configuration though so I just wanted to checkin.

**Comment by amityweb** [comment]
I’ve not had complaints. But make sure you don’t use Cloudflare (or disable caching images if you do) because you can’t use Cloudflare free plan to cache images and serve webp yourself. Cloudflare will serve cached webp to Safari. 

You either disable caching images on Cloudflare or pay to use their Polish service and they create webp for you.

**Comment by amityweb** [comment]
Looks like purelaza.org is using Cloudflare. So that’ll be the issue.

**Comment by mathieu-aubin** [comment]
> Will most likely fail with services such as CloudFlare and heavy-ish caching/page rules settings

Got that right!

**Comment by datvnn** [comment]
Not that the Webp image doesn't work anymore. Webp is still distributed in the browser. When "Enable WebP caching" is enabled in WP Rocket Plugin, Plugin will generate index-https-webp.html and index-https-webp.html_gzip. But x-rocket-nginx-file only accepts index-https.html. So not sure if the cache file is still served correctly?
If using Webp Express, the temporary solution is to deactivate "Enable WebP caching" and use Alter HTML of Webp Express.

**Comment by maximejobin** [comment]
WebP has never worked with Rocket-Nginx. The code is almost completed and will be shipped with version 3.0.

As said earlier in the conversation, we are waiting for a fix from WP Rocket to make it work with Rocket-Nginx.

You can add your comment on the ticket created in October on their GitHub to try to speedup the process.

See the link to find where the conversation is on their GitHub:
https://github.com/SatelliteWP/rocket-nginx/issues/136#issuecomment-709537368

**Comment by maximejobin** [comment]
It was decided that this feature would not be supported.

I truly believe that generating a HTML file specific to WebP is a bad idea. WebP is widely supported and should not get its specific file.

Therefore, this will be closed.
